### DISCUSSION TITLE
Routing between two spokes in "Hub and spoke" configuration

### DISCUSSION THREAD

**Discussion by xzkostyan** [strongswan/strongswan#2486] [Q&A] [answered]
Hi.

I want to connect two gateways (B and C) to central hub (A) like described in "Hub and spoke" https://wiki.strongswan.org/projects/strongswan/wiki/SubnetsBehindMoreThanTwoGateways. Also I want routing between B and C via A.

Central hub A has public IP and network 192.168.0.0/24 behind it.
Gateways B and C doesn't have public IP.
There is network 192.168.1.0/24 behind gateway B and 192.168.3.0/24 behind gateway C. Strongswan is used for IKEv2 in all nodes.

Here is working example for A <-> B and A <-> C routing. I can ping 192.168.1.100 and 192.168.3.100 from machine 192.168.0.100.

```
config setup
    charondebug="ike 2, knl 2, cfg 2"

conn rw-base
    fragmentation=yes
    dpdaction=clear
    dpdtimeout=90s
    dpddelay=300s

сonn ikev2-base
    fragmentation=yes
    dpdaction=clear
    dpdtimeout=90s
    dpddelay=300s

    keyexchange=ikev2
    forceencaps=yes

    ike=...
    esp=...

    mark=%unique
    leftauth=psk
    rightauth=psk

conn ikev2-B
    also=ikev2-base

    leftsubnet=192.168.0.0/24
    rightid=someid
    rightsubnet=192.168.1.0/24
    rightsendcert=never

    auto=add

conn ikev2-C
    also=ikev2-base

    leftsubnet=192.168.0.0/24
    rightid=someid2
    rightsubnet=192.168.3.0/24
    rightsendcert=never

    auto=add
```

When I try to add extra subnets to `leftsubnet`

```
conn ikev2-B
    also=ikev2-base
    leftsubnet=192.168.0.0/24,192.168.3.0/24
    ...

conn ikev2-C
    also=ikev2-base
    leftsubnet=192.168.0.0/24,192.168.1.0/24
    ...
```

I got following:

```
selecting traffic selectors for us:
 config: 192.168.0.0/24, received: 192.168.0.0/24 => match: 192.168.0.0/24
 config: 192.168.0.0/24, received: 192.168.1.0/24 => no match
 config: 192.168.1.0/24, received: 192.168.0.0/24 => no match
 config: 192.168.1.0/24, received: 192.168.1.0/24 => match: 192.168.1.0/24
selecting traffic selectors for other:
 config: 192.168.3.0/24, received: 192.168.3.0/24 => match: 192.168.3.0/24
getting a local address in traffic selector 192.168.0.0/24
using host 192.168.0.1
installing route: 192.168.3.0/24 src 192.168.0.1 dev ipsec0
getting iface index for ipsec0
getting a local address in traffic selector 192.168.1.0/24
no local address found in traffic selector 192.168.1.0/24
error installing route with policy 192.168.1.0/24 === 192.168.3.0/24 out
unable to install IPsec policies (SPD) in kernel
```

Obviously there is no local address 192.168.1.0/24 on hub A.

How can I organize B <-> C routing with pure ipsec? Does OSPF required for this?

I don't think that current routing table and firewall rules on hub A are relevant. But if they are I'll post them.

**Comment by tobiasbrunner** [comment, marked as answer]
Don't use [kernel-libipsec](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) then.

**Comment by xzkostyan** [comment]
Thanks.

Here is more detailed answer.

Without kernel-libipsec you need to manage 'ipsec' interface and routes by yourself.

On hub A you need to:

0) Rewrite config from ipsec.conf to swanctl.conf. XFRM doesn't work with ipsec.conf
1) Choose id for interface: 42 for example.
2) Create link `ip link add ipsec0 type xfrm dev enp9s0 if_id 42`.
3) Add routes `ip route add 192.168.1.0/24 dev ipsec0 proto static src 192.168.0.1`, `ip route add 192.168.3.0/24 dev ipsec0 proto static src 192.168.0.1`.
4) Modify `local_ts` to `local_ts=192.168.0.0/24,192.168.1.0/24`, `local_ts=192.168.0.0/24,192.168.3.0/24` for gateway B and gateway C.
5) Add `if_id_in=42` and `if_id_out=42` into each `children` item section.

https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html

**Reply by tobiasbrunner** [reply]
If you want interfaces, sure. But you don't _need_ any interfaces for this scenario.
